components/telegram_bot_api/actions/list-chats/list-chats.js

```diff
@@ -0,0 +1,72 @@
+/* eslint-disable camelcase */
+const telegram_bot_api = require("../../telegram_bot_api.app.js");
+
+module.exports = {
+  key: "telegram_bot_api-list-chats",
+  name: "List Chats",
+  description: "List available Telegram chats",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    telegram_bot_api,
+    offset: {
+      propDefinition: [
+        telegram_bot_api,
+        "offset",
+      ],
+    },
+    limit: {
+      propDefinition: [
+        telegram_bot_api,
+        "limit",
+      ],
+    },
+    autoPaging: {
+      propDefinition: [
+        telegram_bot_api,
+        "autoPaging",
+      ],
+    },
+  },
+  methods: {
+    /**
+     * Uses an Update object to generate a result object containing chat
+     * metadata, representing the source of the update and chat in which the
+     * update was made
+     *
+     * @param {Object} update - an Update to use to generate a result
+     * @returns a result representing a chat
+     */
+    generateResultFromUpdate(update) {
+      return {
+        update_id: update.update_id,
+        ...Object.keys(update).reduce((resultPart, key) => {
+          return (resultPart.chat
+            ? resultPart
+            : {
+              chat: update[key].chat,
+              from: update[key].from,
+              forward_from_chat: update[key].forward_from_chat,
+            });
+        }, {}),
+      };
+    },
+  },
+  async run() {
+    const updates = await this.telegram_bot_api.getUpdates({
+      offset: this.offset,
+      limit: this.limit,
+    });
+    if (this.autoPaging && updates.length > 0) {
+      const lastUpdateId = updates[updates.length - 1].update_id;
+      // "Confirm" updates by calling API to get updates using an offset of the
+      // last `update_id + 1` See [documentation for getUpdates in the Telegram
+      // Bot API reference](https://core.telegram.org/bots/api#getupdates)
+      await this.telegram_bot_api.getUpdates({
+        offset: lastUpdateId + 1,
+        limit: 1,
+      });
+    }
```

Comment:
I didn't quite get how does this work. I mean you are calling only twice the `getUpdates` function in case `autoPaging` is true and there is still records in the list, but you are just making the request without reassgining the `updates` array? also what happens if there is more than 2 pages? and should you set a default limit for the whole pagination? like 30 records max? same questions go for `list-updates`

Reply:
According to the [Telegram Bot API docs](https://core.telegram.org/bots/api#getupdates):
> By default, updates starting with the earliest unconfirmed update are returned. An update is considered confirmed as soon as getUpdates is called with an offset higher than its update_id.

When `getUpdates` is called with an `offset`, following calls to `getUpdates` _without_ an `offset` will return updates with `update_id` >= (previously called) max `offset`.

- When `autoPaging` is `true`, this action "confirms" the updates by calling `getUpdates` with `offset = lastUpdateId + 1`.
- Each following execution of this action returns only updates that haven't yet been processed (`update_id >= lastOffset`)--the Telegram server handles pagination.

Do you think we should optionally paginate through all of the new updates and not just `limit` updates?

Reply:
Well that was discussed some days ago on slack, but I will say I'll leave it to you. For me it looks good!